### PR TITLE
Zilien parser

### PR DESCRIPTION
**Pull request by KennethNielsen**
This PR contains the implementation of a ixdat native Zilien tsv parser. Besides from a re-implementation, there is some added functionality, as per the change log:

- An ixdat native parser for Spectro Inlets - Zilien files has been
  implemented, as a replacement for leveraging the one in the legacy
  EC_MS project. While re-implementing it, support was also added for
  all auxiliary data series like MFC, PC, ion gauge and pirani values
  etc. Also, the Zilien tsv files can now be loaded as both EC-MS, MS
  and EC measurements, which solves some issues pertaining to
  plotting. To load a Zilien tsv file as something else than the
  default ``ECMSMeasurement``, do either:

    >>> ms_measurement = MSMeasurement.load(path_to_file="...")

  or:

    >>> ms_measurement = Measurement.load(path_to_file="...", technique="MS")

  This added functionality solves
  `Issue #31 <https://github.com/ixdat/ixdat/issues/31>`_

Besides from this, the PR also adds a small Zilien tsv data set and implements high level functional tests for the parser.

What I'm especially interested in, in a review, is whether all the correct information is being passed into the Measurements at the end of the read method.

### CODE REVIEW

File: src/ixdat/readers/zilien.py [hunk 1]
```diff
@@ -1,61 +1,314 @@
...
+from matplotlib import pyplot as plt
+
 from ..data_series import DataSeries, TimeSeries, ValueSeries, Field
 from ..techniques import ECMSMeasurement, MSMeasurement, ECMeasurement, Measurement
 from ..techniques.ms import MSSpectrum
 from .reading_tools import timestamp_string_to_tstamp, FLOAT_MATCH
-from .ec_ms_pkl import measurement_from_ec_ms_dataset
+
 
 ZILIEN_TIMESTAMP_FORM = "%Y-%m-%d %H_%M_%S"  # like 2021-03-15 18_50_10
```
Comment (ScottSoren) [inline src/ixdat/readers/zilien.py:19]: Zilien files get saved with a different timestamp form. For example: "2022_03_30_18_47_24_my_title.tsv"**
When that happens here I've asked my colleagues to change it to the normal timestamp format. But it would be nice if they didn't have to.

File: src/ixdat/readers/zilien.py [hunk 2]
```diff
@@ -186,15 +439,20 @@ def read(self, path_to_spectrum, cls=None, **kwargs):
...
-        "zilien_with_ec/2021-02-01 17_44_12.tsv"
+    path_to_test_file = (
+        Path(__file__).parent.resolve().parent.parent.parent
+        / "test_data"
+        / "Zilien version 1"
+        / "2022-04-06 16_17_23 full set.tsv"
     )
 
-    ecms_measurement = Measurement.read(
+    ecms_measurement = ECMSMeasurement.read(
```
Comment (ScottSoren) [inline src/ixdat/readers/zilien.py:449]: why the change here?
`Measurement.read()` will be the default way to start in ixdat, and should do the right thing as much as possible!

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt [hunk 3]
```diff
@@ -0,0 +1,938 @@
+EC-Lab ASCII FILE
```
Comment (ScottSoren) [inline test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt:1]: Likewise, I wonder if it's worth thinning out the data here like in Pt_poly_cv_CUT.mpt?

File: tests/functional/test_zilien_parser.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+import datetime
+import time
+from collections import namedtuple
+from pathlib import Path
+
+import pytest
+from pytest import approx
+from ixdat.techniques.ec_ms import ECMSMeasurement, MSMeasurement, ECMeasurement
+
+DATA_DIR = Path(__file__).parent / ".." / ".." / "test_data"
```
Comment (ScottSoren) [inline tests/functional/test_zilien_parser.py:12]: why not `Path(__file__).parent.parent.parent / "test_data"`?

File: NEXT_CHANGES.rst [hunk 5]
```diff
@@ -13,3 +13,29 @@ links to relevant Issues, Discussions, and PR's on github with the following for
 `PR #XX <https://github.com/ixdat/ixdat/pulls/XX>`_
 
 
+ixdat next
+========================
+
+changed
+-------
+
+parsers
```
Comment (ScottSoren) [inline NEXT_CHANGES.rst:22]: "readers"

File: tests/functional/test_zilien_parser.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    ),
+}
+
+
+@pytest.mark.parametrize(
+    ["cls", "expected_series"],
+    (
+        (ECMSMeasurement, EXPECTED_SERIES_ECMS),
+        (MSMeasurement, EXPECTED_SERIES_MS),
+        (ECMeasurement, EXPECTED_SERIES_EC),
```
Comment (ScottSoren) [inline tests/functional/test_zilien_parser.py:119]: Can you add `(Measurement, EXPECTED_SERIES_ECMS)` here? This will be the most common way to invoke the reader.

File: tests/functional/test_zilien_parser.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    ),
+)
+def test_read(cls, expected_series):
+    """Test parsing the file as an ECMSMeasurement"""
+    measurement = cls.read(PATH_TO_DATAFILE, reader="zilien")
+    assert type(measurement) == cls
```
Comment (ScottSoren) [inline tests/functional/test_zilien_parser.py:125]: this is not the case for `cls = Measurement`

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv [hunk 8]
```diff
@@ -0,0 +1,850 @@
+num_header_lines	Number of header lines		int	115
```
Comment (ScottSoren) [inline test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv:1]: I wonder if it's worth shrinking the data in the file down, like you did for test_data/biologic/Pt_poly_cv_CUT.mpt?

File: tests/functional/test_zilien_parser.py (hunk 6 above)
Comment (ScottSoren) [reply tests/functional/test_zilien_parser.py:119]: And likewise, 
`Measurement.read(PATH_TO_DATAFILE, reader="zilien", technique="MS")` should give it the `EXPECTED_SERIES_MS`, etc.

File: tests/functional/test_zilien_parser.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    timestamp = time.mktime(
+        datetime.datetime.strptime(timestamp_string, TIME_FORMAT).timetuple()
+    )
+    assert measurement.tstamp == approx(timestamp)
+
+    # Test series and series data properties
+    for series_name in expected_series:
+        if series_name.endswith("[s]"):
+            continue
+        single_series_details = SERIES_DETAILS[series_name]
```
Comment (ScottSoren) [inline tests/functional/test_zilien_parser.py:139]: nice.

**Review by ScottSoren — CHANGES_REQUESTED** [review]
Hi Kenneth,

Nice reader!
Most comments here are not big deals. The main thing is that this has to work:
`ms = Measurement.read(PATH_TO_DATA, reader="zilien", technique="MS")`
I think it's just adding a `measurement_kwargs.update(kwargs)`.

File: src/ixdat/readers/zilien.py [hunk 10]
```diff
@@ -1,61 +1,314 @@
...
+        aliases = dict(aliases)  # Convert from defaultdict to normal dict
+
+        measurement_kwargs = {
+            "name": name or file_stem,
+            "series_list": series,
+            "aliases": aliases,
+            "tstamp": timestamp,
+            "metadata": metadata,
+        }
+        self._measurement = cls(**measurement_kwargs)
```
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:175]: Hmm. I agree that the update line is missing. Nice catch. I'm a little puzzled, because it hasn't failed for me. In any case, I have added a test case for each of the three options of read them from `Measurement` and with differing techniques.

File: src/ixdat/readers/zilien.py [hunk 11]
```diff
@@ -1,61 +1,314 @@
...
                 kwargs["technique"] = "EC-MS"
             elif issubclass(cls, ECMeasurement):
                 kwargs["technique"] = "EC"
             elif issubclass(cls, MSMeasurement):
                 kwargs["technique"] = "MS"
+        self._cls = cls
+
+        self._path_to_file = Path(path_to_file)
+
+        # Extra timestamp from filename on form: 2021-04-20 11_16_18 Measurement name
```
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:139]: Didn't realize it was there, thanks

File: src/ixdat/readers/zilien.py [hunk 12]
```diff
@@ -1,61 +1,314 @@
...
+        """Read metadata from `file_handle`"""
+        metadata = {}
+        # The first lines always include 3 pieces of information about the length of the
+        # header and sometimes a file format version. Start by reading the first 3 lines,
+        # to figure out if the version is amongst them
+        next_line_number_to_read = 3
+        for n in range(next_line_number_to_read):
+            key, value = parse_metadata_line(file_handle.readline())
+            metadata[key] = value
+
```
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:296]: Good point, added

File: src/ixdat/readers/zilien.py [hunk 13]
```diff
@@ -1,61 +1,314 @@
...
+        Returns:
+            str, str, Optional[str]: Return series_name, unit, standard_name
+
+        """
+        standard_name = None
+        if column_header == "Time [s]":  # Form TimeSeries
+            unit = "s"
+            if series_header == "pot":
+                name = f"Potential {column_header.lower()}"
```
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:254]: They could. But since they will no depend on the content of the file, I think I would prefer to keep the there for now. Will change that, when there is a zilien file format version 2

File: src/ixdat/readers/zilien.py (hunk 1 above)
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:19]: Agreed. What's your opinion on this, should I try and account for it here, or merely log it as a olefn issue (which are now finally being worked on) ?

File: src/ixdat/readers/zilien.py (hunk 2 above)
Comment (KennethNielsen) [reply src/ixdat/readers/zilien.py:449]: Agreed. I think this was just for testing. I will adjust it and put it in a function etc.

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt (hunk 3 above)
Comment (KennethNielsen) [reply test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt:1]: Maybe. The thing is, that the biologic ones are not even used for now, so maybe I should just delete them and delete them from the history. What do you think?

It's another story about the tsv, see below.

File: tests/functional/test_zilien_parser.py (hunk 4 above)
Comment (KennethNielsen) [reply tests/functional/test_zilien_parser.py:12]: I really don't know :D Fixed

File: NEXT_CHANGES.rst (hunk 5 above)
Comment (KennethNielsen) [reply NEXT_CHANGES.rst:22]: That is probably prevalent. I will grep try and grep the diff of the entire PR to look for it.

File: tests/functional/test_zilien_parser.py (hunk 6 above)
Comment (KennethNielsen) [reply tests/functional/test_zilien_parser.py:119]: Both added along with `technique="EC"` and friends. There are now 7 test in total:
```python
    (
        # cls_or_technique, expected_series
        (Measurement, EXPECTED_SERIES_ECMS),
        (ECMSMeasurement, EXPECTED_SERIES_ECMS),
        (MSMeasurement, EXPECTED_SERIES_MS),
        (ECMeasurement, EXPECTED_SERIES_EC),
        ("EC-MS", EXPECTED_SERIES_ECMS),
        ("EC", EXPECTED_SERIES_EC),
        ("MS", EXPECTED_SERIES_MS),
    ),
```
with the test case taking care of determining which class the result should be:
```python
    if isinstance(cls_or_technique, str):
        expected_measurement_class = TECHNIQUE_NAME_TO_CLASS[cls_or_technique]
        measurement = Measurement.read(
            PATH_TO_DATAFILE, reader="zilien", technique=cls_or_technique
        )
    else:
        measurement = cls_or_technique.read(PATH_TO_DATAFILE, reader="zilien")
        if cls_or_technique is Measurement:
            expected_measurement_class = ECMSMeasurement
        else:
            expected_measurement_class = cls_or_technique

    assert type(measurement) == expected_measurement_class
```

File: tests/functional/test_zilien_parser.py (hunk 7 above)
Comment (KennethNielsen) [reply tests/functional/test_zilien_parser.py:125]: Taken care, see the other comment about test cases.

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv (hunk 8 above)
Comment (KennethNielsen) [reply test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv:1]: The thing is, when I cut the other one, it was out of concern for test time, but since all the data is now being read in through numpy.genfromtxt I don't think it will make much difference. I decided to measure it and the data supports it:
```text
…/ixdat zilien_parser +50-26 $» v3.9.5 (ixdat) 
at 11:41:28 ❯ pytest tests -k zilien_reader -vv --durations=0
================================================= test session starts ==================================================
platform linux -- Python 3.9.5, pytest-7.1.0, pluggy-1.0.0 -- /home/kenneth/venv/ixdat/bin/python
cachedir: .pytest_cache
rootdir: /home/kenneth/venv/ixdat/ixdat
plugins: profiling-1.7.0
collecting ... importing ixdat v0.2.1 from /home/kenneth/venv/ixdat/ixdat/src/ixdat/__init__.py
collected 93 items / 86 deselected / 7 selected                                                                        

tests/functional/test_zilien_reader.py::test_read[Measurement-expected_series0] PASSED                           [ 14%]
tests/functional/test_zilien_reader.py::test_read[ECMSMeasurement-expected_series1] PASSED                       [ 28%]
tests/functional/test_zilien_reader.py::test_read[MSMeasurement-expected_series2] PASSED                         [ 42%]
tests/functional/test_zilien_reader.py::test_read[ECMeasurement-expected_series3] PASSED                         [ 57%]
tests/functional/test_zilien_reader.py::test_read[EC-MS-expected_series4] PASSED                                 [ 71%]
tests/functional/test_zilien_reader.py::test_read[EC-expected_series5] PASSED                                    [ 85%]
tests/functional/test_zilien_reader.py::test_read[MS-expected_series6] PASSED                                    [100%]

================================================== slowest durations ===================================================
0.03s call     tests/functional/test_zilien_reader.py::test_read[Measurement-expected_series0]
0.01s call     tests/functional/test_zilien_reader.py::test_read[ECMSMeasurement-expected_series1]
0.01s call     tests/functional/test_zilien_reader.py::test_read[EC-expected_series5]
0.01s call     tests/functional/test_zilien_reader.py::test_read[EC-MS-expected_series4]
0.01s call     tests/functional/test_zilien_reader.py::test_read[MSMeasurement-expected_series2]
0.01s call     tests/functional/test_zilien_reader.py::test_read[MS-expected_series6]
0.01s call     tests/functional/test_zilien_reader.py::test_read[ECMeasurement-expected_series3]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[Measurement-expected_series0]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[ECMeasurement-expected_series3]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[EC-expected_series5]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[ECMSMeasurement-expected_series1]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[MS-expected_series6]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[EC-MS-expected_series4]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[MSMeasurement-expected_series2]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[ECMeasurement-expected_series3]
0.00s setup    tests/functional/test_zilien_reader.py::test_read[EC-expected_series5]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[Measurement-expected_series0]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[ECMSMeasurement-expected_series1]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[MSMeasurement-expected_series2]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[MS-expected_series6]
0.00s teardown tests/functional/test_zilien_reader.py::test_read[EC-MS-expected_series4]
=========================================== 7 passed, 86 deselected in 0.89s ===========================================
```
As you can see, actually executing all these tests amount to just 100ms. The rest of the 0.9s must be spent discovering the tests. The only other concern is archive size (on disk and in transfer), but here I think ~500kb per parser on disk and 115kb compressed in transfer is not an unreasonable amount, especially considering we will likely not have hundreds of parsers.

File: tests/functional/test_zilien_parser.py (hunk 9 above)
Comment (KennethNielsen) [reply tests/functional/test_zilien_parser.py:139]: I'm glad you like it :D

File: src/ixdat/readers/zilien.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ScottSoren) [reply src/ixdat/readers/zilien.py:170]: It worked when you import the class, as you do in the test. With this update it also work when the technique is passed as a kwarg.

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv (hunk 8 above)
Comment (ScottSoren) [reply test_data/Zilien version 1/2022-04-06 16_17_23 full set.tsv:1]: got it, then it's fine :)

File: test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt (hunk 3 above)
Comment (ScottSoren) [reply test_data/Zilien version 1/2022-04-06 16_17_23 full set_01_02_CVA_DUSB0_C01.mpt:1]: It's not a big deal with space, as you said, and maybe someone will be interested in them for writing other tests. I think fine to leave in the repository as is.

File: tests/functional/test_zilien_parser.py (hunk 6 above)
Comment (ScottSoren) [reply tests/functional/test_zilien_parser.py]: nice!
:) 
It looks like all is tested.

File: src/ixdat/readers/zilien.py (hunk 1 above)
Comment (ScottSoren) [reply src/ixdat/readers/zilien.py:17]: I think it seems to only happen when the measurement has exited in an improper way.  So... fine to leave to leave it as is here.
Otherwise it'd just require inputting `forms` instead of `form` on line 141. I may do so in a patch if it comes up again here (hopefully not!).

**Review by ScottSoren — APPROVED** [review]
Looks good with the changes! Ready to merge :)
